enhancements/kube-apiserver/audit-policy.md

```diff
@@ -0,0 +1,494 @@
+we hesitate to go the same route: if the file-based API goes away, it is infeasible to patch it back into our OpenShift API server
+binaries without major effort.
+
+From the requirements we get from customers, it is apparent that the file-based policies are far too low-level, and
+too complex to get right. This enhancement is about this high-level API which we can offer to satisfy the needs of
+customers with strong regulatory or security requirement, while being easy to configure for them and feasible to 
+maintain in the future.
+
+This section is for explicitly listing the motivation, goals and non-goals of
+this proposal. Describe why the change is important and the benefits to users.
+
+### Goals
+
+- add an abstract audit policy configuration to `apiservers.config.openshift.io/v1`
+  - with few predefined policies
+  - easily extensible in the future 
+  - few knobs on the level of concrete OpenShift customer requirement
+- applied by kube-apiserver, openshift-apiserver and oauth-apiserver.
+
+### Non-Goals
+
+- configuration of audit backends like webhooks
+- configuration of file logger parameters
+
+## Proposal
+
+We propose to add a `audit.profile` field to `apiservers.config.openshift.io/v1` which selects the 
+audit policy to be deployed to all OpenShift-provided APIservers in the cluster. From the beginning
+we provide the following profile:
+
+- `Default`: this is the existing default policy. `Default` is also the default value in the `APIServer` custom resource:
+
+  ```yaml
+    apiVersion: audit.k8s.io/v1beta1
+    kind: Policy
+    # Don't generate audit events for all requests in RequestReceived stage.
+    omitStages:
+    - "RequestReceived"
+    rules:
+    # Don't log requests for events
+    - level: None
+      resources:
+      - group: ""
+        resources: ["events"]
+    # Don't log oauth tokens as metadata.name is the secret
+    - level: None
+      resources:
+      - group: "oauth.openshift.io"
+        resources: ["oauthaccesstokens", "oauthauthorizetokens"]
+    # Don't log authenticated requests to certain non-resource URL paths.
+    - level: None
+      userGroups: ["system:authenticated", "system:unauthenticated"]
+      nonResourceURLs:
+      - "/api*" # Wildcard matching.
+      - "/version"
+      - "/healthz"
+      - "/readyz"
+  
+    # Here the tail begins which differs per profile:
+  
+    # A catch-all rule to log all other requests at the Metadata level.
+    - level: Metadata
+      # Long-running requests like watches that fall under this rule will not
+      # generate an audit event in RequestReceived.
+      omitStages:
+      - "RequestReceived"
+   ```
+
+- `OauthUserRequests`: this is like `Default`, but it logs request bodies for requests for oauth users only, i.e. by users outside of the cluster.
+- `NonOpenShiftRequests`: this is like `Default`, but it logs request bodies for requests not coming from known OpenShift components like kubelets, controllers and operators.
```

Comment:
I don't see value here.  This could purely be done with post filtering.

Reply:
post-filtering means IO and cpu load. If we don't filter at all, we need numbers from perf team about the overhead.

Reply:
> post-filtering means IO and cpu load. If we don't filter at all, we need numbers from perf team about the overhead.

if a customer truly has a requirement to record everything, then they won't care how much it costs and they won't be willing to simply ignore "special" actors, because the all the mutators matter.  Especially since anyone who can see those creds and use those creds.

Reply:
@jtaleric we need numbers for cpu and IO overhead when we enable a deep audit policy. Would you be able to measure that if I give you a policy and some instruction how to enable that?

Reply:
> @jtaleric we need numbers for cpu and IO overhead when we enable a deep audit policy. Would you be able to measure that if I give you a policy and some instruction how to enable that?

## System Under Test
**Platform :** AWS
**OCP Version :** 4.4.0-rc.2
**Kubernetes Version:** v1.17.1
**Worker Node count:** 250
**Master :** 3
**Infrastructure :** 3
**Masters Nodes :** r5.4xlarge
**Infrastructure Nodes :** m5.12xlarge
**Worker Nodes :**  m5.2xlarge

## 256 Nodes Idle
 Policy | CPU | Memory | Disk 
----------|--------|-------------|-------
Baseline | 11.93% | 1.66GiB | 617kBs
Write RequestResponse | 15.34% | 1.607GiB | 570kBs 
Write Request | 21.55% | 1.792GiB | 654kBs
R+W Request Response | 20.39% | 2.035GiB | 880kBs

## 256 Nodes Under load
 Policy | CPU | Memory | Disk 
----------|--------|-------------|-------
Baseline | 39% | 2.154GiB | 1.277MBs
Write RequestResponse | 32% | 2.085GiB | 1.3MBs 
Write Request | 45% | 2.397GiB | 1.448MBs
R+W Request Response | 47% | 2.393GiB | 1.974MBs

The load was running clusterloader - master vertical testing with 100 projects. 

The CPU and memory were captured querying 
- `﻿container_cpu_usage_seconds_total` looking at the openshift-kube-apiserver namespace.
- `container_memory_rss` looking at the openshift-kube-apiserver namespace.

For the Disk 
- `﻿node_disk_read_bytes_total` 

Hope this helps!

Reply:
> Hope this helps!

It does!